File: src/jfan/fan/sys/NullableType.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   7 Oct 08  Brian Frank  Creation
//
package fan.sys;

/**
 * NullableType wraps a type as nullable with trailing "?".
 */
public class NullableType
  extends Type
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  NullableType(Type root)
  {
    this.root = root;
    this.signature = root.signature() + "?";
  }

//////////////////////////////////////////////////////////////////////////
// Type
//////////////////////////////////////////////////////////////////////////

  public Pod pod() { return root.pod(); }
  public String name() { return root.name(); }
  public String qname() { return root.qname(); }
  public String signature() { return signature; }
  int flags() { return root.flags(); }

  public Type base() { return root.base(); }
  public List mixins() { return root.mixins(); }
  public List inheritance() { return root.inheritance(); }
  public boolean is(Type type) { return root.is(type); }

  public Boolean isNullable() { return true; }
  protected Type makeToNullable() { return this; }

  protected Type makeToListOf() { return new ListType(root).toNullable(); }

  public List fields() { return root.fields(); }
  public List methods() { return root.methods(); }
  public List slots() { return root.slots(); }
  public Slot slot(String name, boolean checked) { return root.slot(name, checked); }

  public Map facets(Boolean inherited) { return root.facets(inherited); }
  public Object facet(String name, Object def, Boolean inherited) { return root.facet(name, def, inherited); }

  public String doc() { return root.doc(); }

  public boolean javaRepr() { return root.javaRepr(); }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  final Type root;
  final String signature;
  private Type listOf;

}

File: src/nfan/fan/sys/NullableType.cs
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   21 Oct 08  Andy Frank  Creation
//

using System.Runtime.CompilerServices;

namespace Fan.Sys
{
  /// <summary>
  /// NullableType wraps a type as nullable with trailing "?".
  /// </summary>
  public class NullableType : Type
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructor
  //////////////////////////////////////////////////////////////////////////

    internal NullableType(Type root)
    {
      m_root = root;
      m_signature = root.signature() + "?";
    }

  //////////////////////////////////////////////////////////////////////////
  // Type
  //////////////////////////////////////////////////////////////////////////

    public override Pod pod() { return m_root.pod(); }
    public override string name() { return m_root.name(); }
    public override string qname() { return m_root.qname(); }
    public override string signature() { return m_signature; }
    internal override int flags() { return m_root.flags(); }

    public override Type @base() { return m_root.@base(); }
    public override List mixins() { return m_root.mixins(); }
    public override List inheritance() { return m_root.inheritance(); }
    public override bool @is(Type type) { return m_root.@is(type); }

    public override Boolean isNullable() { return Boolean.True; }
    public override Type toNullable() { return this; }
    public override Type toNonNullable() { return m_root; }

    public override bool isGenericType() { return m_root.isGenericType(); }
    public override bool isGenericInstance() { return m_root.isGenericInstance(); }
    public override bool isGenericParameter() { return m_root.isGenericParameter(); }
    public override Type getRawType() { return m_root.getRawType(); }
    public override Map @params() { return m_root.@params(); }
    public override Type parameterize(Map pars) { return m_root.parameterize(pars).toNullable(); }

    public override Boolean isDynamic() { return Boolean.False; }

    public override List fields() { return m_root.fields(); }
    public override List methods() { return m_root.methods(); }
    public override List slots() { return m_root.slots(); }
    public override Slot slot(string name, bool check) { return m_root.slot(name, check); }

    public override Map facets(Boolean inherited) { return m_root.facets(inherited); }
    public override object facet(string name, object def, Boolean inherited) { return m_root.facet(name, def, inherited); }

    public override string doc() { return m_root.doc(); }

    public override bool netRepr() { return m_root.netRepr(); }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal readonly Type m_root;
    internal readonly string m_signature;

  }
}